services/librariesio/librariesio-api-provider.js

```diff
@@ -0,0 +1,106 @@
+    }
+  }
+
+  getRateLimitFromHeaders({ headers, token }) {
+    // The Libraries.io API does not consistently provide the rate limiting headers.
+    // In some cases (e.g. package/version not founds) it won't include any of these headers,
+    // and the `retry-after` header is only provided _after_ the rate limit has been exceeded
+    // and requests are throttled.
+    //
+    // https://github.com/librariesio/libraries.io/issues/2860
+
+    // The standard rate limit is 60/requests/minute, so fallback to that default
+    // if the header isn't present.
+    // https://libraries.io/api#rate-limit
+    const rateLimit = headers['x-ratelimit-limit'] || this.defaultRateLimit
+
+    // If the remaining header is missing, then we're in the 404 response phase, and simply
+    // subtract one from the `usesRemaining` count on the token, since the 404 responses do count
+    // against the rate limits.
+    const totalUsesRemaining =
+      headers['x-ratelimit-remaining'] || token.decrementedUsesRemaining
+
+    // The `retry-after` header is only present post-rate limit excess, and contains the value in
+    // seconds the client needs to wait before the limits are reset.
+    // Our token pools internally use UTC-based milliseconds, so we perform the conversion
+    // if the header is present to ensure the token pool has the correct value.
+    // If the header is absent, we just use the current timestamp to
+    // advance the value to _something_
+    const retryAfter = headers['retry-after']
+    const nextReset = Date.now() + (retryAfter ? retryAfter * 1000 : 0)
+
+    return {
+      rateLimit,
+      totalUsesRemaining,
+      nextReset,
+    }
+  }
+
+  updateToken({ token, res }) {
+    const { totalUsesRemaining, nextReset } = this.getRateLimitFromHeaders({
+      headers: res.headers,
+      token,
+    })
+    token.update(totalUsesRemaining, nextReset)
+  }
+
+  async requestAsPromise(request, url, options = {}) {
+    const { baseUrl } = this
+
+    let token
+    let tokenString
+    if (this.withPooling) {
+      try {
+        token = this.standardTokens.next()
+      } catch (e) {
+        console.error('Unable to select next Libraries.io token from pool')
+        log.error(e)
+        return
+      }
+      tokenString = token.id
+    } else {
+      tokenString = this.globalToken
+    }
+
+    const mergedOptions = {
+      ...options,
+      ...{
+        headers: {
+          'User-Agent': userAgent,
+          ...options.headers,
+        },
+        qs: {
+          api_key: tokenString,
+          ...options.qs,
+        },
+      },
+    }
+    const response = await request(`${baseUrl}${url}`, mergedOptions)
+    if (this.withPooling) {
+      if (response.res.statusCode === 401) {
+        this.invalidateToken(token)
+      } else if (response.res.statusCode < 500) {
+        this.updateToken({ token, url, res: response.res })
+      }
+    }
+    return response
```

Comment:
My reading of how we've set up the actual http request issuing code with got is that this will only reject/throw in the event of major issue (no connectivity) and not on standard http response errors. In such a major issue type of scenario, I don't think it would be safe/valid to attempt to update the token pooling info, so I've not wrapped this in any kind of try catch. I _think_ the call making infrastructure already handles anything that would be caught here, but let me know if a catch is needed with something like an explicit error conversion